prov/shm/src/smr_progress.c

```diff
@@ -202,8 +217,93 @@ static int smr_progress_iov(struct smr_cmd *cmd, struct iovec *iov,
 	return -ret;
 }
 
+static int smr_mmap_iov_copy_out(struct smr_ep *ep, struct smr_cmd *cmd,
+				 struct iovec *iov, size_t iov_count,
+				 size_t *total_len)
+{
+	char shm_name[NAME_MAX];
+	char peer_name[NAME_MAX];
+	void *mapped_ptr;
+	int peer_id, fd, ret, num;
+
+	peer_id = (int) cmd->msg.hdr.addr;
+
+	sprintf(peer_name, "%s", ep->region->map->peers[peer_id].peer.name);
+	num = sprintf(shm_name, "%s_%s_%ld", peer_name, ep->name,
+		      cmd->msg.hdr.msg_id);
+	if (num < 0) {
+		FI_WARN(&smr_prov, FI_LOG_AV, "generating shm file name failed\n");
+		return errno;
+	}
+	assert(num <= NAME_MAX);
+
+	fd = shm_open(shm_name, O_RDWR, S_IRUSR | S_IWUSR);
+	if (fd < 0) {
+		FI_WARN(&smr_prov, FI_LOG_AV, "shm_open error\n");
+		return errno;
+	}
+
+	mapped_ptr = mmap(NULL, cmd->msg.hdr.size, PROT_READ | PROT_WRITE,
+			  MAP_SHARED, fd, 0);
+	if (mapped_ptr == MAP_FAILED) {
+		FI_WARN(&smr_prov, FI_LOG_AV, "mmap error %s\n", strerror(errno));
+		return errno;
+	}
+	if (cmd->msg.hdr.op == ofi_op_read_req) {
+		*total_len = ofi_total_iov_len(iov, iov_count);
+		ret = ofi_copy_from_iov(mapped_ptr, *total_len, iov, iov_count, 0);
+		if (ret != *total_len) {
+			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+				"mmap iov copy in error\n");
+			return FI_EIO;
+		}
+	} else {
+		ret = ofi_copy_to_iov(iov, iov_count, 0, mapped_ptr,
+				      cmd->msg.hdr.size);
+		if (ret != cmd->msg.hdr.size) {
+			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+				"mmap iov copy out error\n");
+			return FI_EIO;
```

Comment:
On error here, do we need to unlink and unmap?

Reply:
Thank you. Fixed.

Reply:
error return codes should be negative for consistency with the rest of the code base.

Reply:
We need to fix the function name.  Are we copying into the mmap region or out of the mmap region, and into or out of the iov?

Reply:
For send-recv, and RMA write, it's mmap region->iov, while for RMA read, it's iov->mmap region. So I'm not sure what the good function name would be. Do you have any suggestion?

Reply:
I renamed it to smr_mmap_peer_copy(), please let me know if it looks good to you.